pdata/internal/grpcencoding/encoding.go

```diff
@@ -0,0 +1,104 @@
+func DefaultBufferPool() mem.BufferPool {
+	return otelBufferPool
+}
+
+// Name is the name registered for the proto compressor.
+const Name = "proto"
+
+func init() {
+	encoding.RegisterCodecV2(&codecV2{delegate: encoding.GetCodecV2(Name)})
+}
+
+// codecV2 is a custom proto encoding that uses a different tier schema for the TieredBufferPool as well
+// as it call into the custom marshal/unmarshal logic that works with memory pooling.
+// If not an otlp payload fallback on the default grpc/proto encoding.
+type codecV2 struct {
+	delegate encoding.CodecV2
+}
+
+func (c *codecV2) Marshal(v any) (mem.BufferSlice, error) {
+	switch req := v.(type) {
+	case *otlpcollectorlogs.ExportLogsServiceRequest:
+		size := internal.SizeProtoOrigExportLogsServiceRequest(req)
+		buf := otelBufferPool.Get(size)
+		n := internal.MarshalProtoOrigExportLogsServiceRequest(req, (*buf)[:size])
+		*buf = (*buf)[:n]
+		return []mem.Buffer{mem.NewBuffer(buf, otelBufferPool)}, nil
+	case *otlpcollectormetrics.ExportMetricsServiceRequest:
+		size := internal.SizeProtoOrigExportMetricsServiceRequest(req)
+		buf := otelBufferPool.Get(size)
+		n := internal.MarshalProtoOrigExportMetricsServiceRequest(req, (*buf)[:size])
+		*buf = (*buf)[:n]
+		return []mem.Buffer{mem.NewBuffer(buf, otelBufferPool)}, nil
+	case *otlpcollectortraces.ExportTraceServiceRequest:
+		size := internal.SizeProtoOrigExportTraceServiceRequest(req)
+		buf := otelBufferPool.Get(size)
+		n := internal.MarshalProtoOrigExportTraceServiceRequest(req, (*buf)[:size])
+		*buf = (*buf)[:n]
+		return []mem.Buffer{mem.NewBuffer(buf, otelBufferPool)}, nil
+	case *otlpcollectorprofile.ExportProfilesServiceRequest:
+		size := internal.SizeProtoOrigExportProfilesServiceRequest(req)
+		buf := otelBufferPool.Get(size)
+		n := internal.MarshalProtoOrigExportProfilesServiceRequest(req, (*buf)[:size])
+		*buf = (*buf)[:n]
+		return []mem.Buffer{mem.NewBuffer(buf, otelBufferPool)}, nil
+	}
+
+	return c.delegate.Marshal(v)
+}
+
+func (c *codecV2) Unmarshal(data mem.BufferSlice, v any) (err error) {
+	switch req := v.(type) {
+	case *otlpcollectorlogs.ExportLogsServiceRequest:
+		// TODO: Upgrade custom Unmarshal logic to support reading from mem.BufferSlice.
+		buf := data.MaterializeToBuffer(otelBufferPool)
+		defer buf.Free()
+		return internal.UnmarshalProtoOrigExportLogsServiceRequest(req, buf.ReadOnlyData())
+	case *otlpcollectormetrics.ExportMetricsServiceRequest:
+		// TODO: Upgrade custom Unmarshal logic to support reading from mem.BufferSlice.
+		buf := data.MaterializeToBuffer(otelBufferPool)
+		defer buf.Free()
+		return internal.UnmarshalProtoOrigExportMetricsServiceRequest(req, buf.ReadOnlyData())
+	case *otlpcollectortraces.ExportTraceServiceRequest:
+		// TODO: Upgrade custom Unmarshal logic to support reading from mem.BufferSlice.
+		buf := data.MaterializeToBuffer(otelBufferPool)
+		defer buf.Free()
+		return internal.UnmarshalProtoOrigExportTraceServiceRequest(req, buf.ReadOnlyData())
```

Comment:
Is the profiling missed intentionally here?

Reply:
Thanks for catching, legit mistake...